// @marblejs/core/dist/+internal/utils/array.util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.insertIfElse = exports.insertIf = exports.mapArray = exports.filterArray = exports.getLast = exports.getHead = exports.getArrayFromEnum = void 0;
const Option_1 = require("fp-ts/lib/Option");
const getArrayFromEnum = (E) => Object.keys(E).filter(key => typeof E[key] === 'number');
exports.getArrayFromEnum = getArrayFromEnum;
const getHead = (array) => (0, Option_1.fromNullable)(array[0]);
exports.getHead = getHead;
const getLast = (array) => (0, Option_1.fromNullable)(array[array.length - 1]);
exports.getLast = getLast;
const filterArray = (f) => (array) => array.filter(f);
exports.filterArray = filterArray;
const mapArray = (f) => (array) => array.map(f);
exports.mapArray = mapArray;
const insertIf = (condition) => (...elements) => condition ? elements : [];
exports.insertIf = insertIf;
const insertIfElse = (condition) => (...elements) => (...elseElements) => condition ? elements : elseElements;
exports.insertIfElse = insertIfElse;
