// @marblejs/core/dist/+internal/utils/string.util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.maskUriComponent = exports.createUuid = exports.stringify = exports.trunc = exports.trim = exports.isString = void 0;
const uuid_1 = require("uuid");
const function_1 = require("fp-ts/lib/function");
const Option_1 = require("fp-ts/lib/Option");
const isString = (value) => typeof value === 'string' || value instanceof String;
exports.isString = isString;
const trim = (strings, ...values) => {
    const notNilValues = values.map((0, function_1.flow)(Option_1.fromNullable, (0, Option_1.getOrElse)(() => '')));
    const interpolation = strings.reduce((prev, current) => prev + current + (notNilValues.length ? notNilValues.shift() : ''), '');
    return interpolation.trim();
};
exports.trim = trim;
const trunc = (n) => (input) => (input.length > n)
    ? input.substr(0, n) + '…'
    : input;
exports.trunc = trunc;
const stringify = (value) => typeof value === 'function'
    ? (value.displayName || value.name)
    : JSON.stringify(value);
exports.stringify = stringify;
const createUuid = () => (0, uuid_1.v4)();
exports.createUuid = createUuid;
const maskUriComponent = (type) => (uri) => {
    switch (type) {
        case 'authorization':
            return uri.replace(/\/\/(.*)\@/, '//[user]:[pass]@');
        default:
            return uri;
    }
};
exports.maskUriComponent = maskUriComponent;
