// @marblejs/core/dist/error/error.model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isEventError = exports.isCoreError = exports.EventError = exports.ContextError = exports.CoreError = exports.ErrorType = void 0;
const utils_1 = require("../+internal/utils");
var ErrorType;
(function (ErrorType) {
    ErrorType["CORE_ERROR"] = "CoreError";
    ErrorType["CONTEXT_ERROR"] = "ContextError";
    ErrorType["EVENT_ERROR"] = "EventError";
})(ErrorType = exports.ErrorType || (exports.ErrorType = {}));
class CoreError extends utils_1.NamedError {
    constructor(message, options) {
        super(ErrorType.CORE_ERROR, message);
        this.message = message;
        Error.prepareStackTrace = (_, stack) => options.stackTraceFactory(message, stack);
        Error.captureStackTrace(this, options.context);
    }
}
exports.CoreError = CoreError;
class ContextError extends utils_1.NamedError {
    constructor(message) {
        super(ErrorType.CONTEXT_ERROR, message);
        this.message = message;
    }
}
exports.ContextError = ContextError;
class EventError extends utils_1.NamedError {
    constructor(event, message, data) {
        super(ErrorType.EVENT_ERROR, message);
        this.event = event;
        this.message = message;
        this.data = data;
    }
}
exports.EventError = EventError;
const isCoreError = (error) => (error === null || error === void 0 ? void 0 : error.name) === ErrorType.CORE_ERROR;
exports.isCoreError = isCoreError;
const isEventError = (error) => (error === null || error === void 0 ? void 0 : error.name) === ErrorType.EVENT_ERROR;
exports.isEventError = isEventError;
