// @marblejs/middleware-logger/dist/logger.util.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.factorizeTime = exports.getTimeDifferenceInMs = exports.formatTime = exports.filterResponse = exports.isNotSilent = exports.getDateFromTimestamp = void 0;
const O = require("fp-ts/lib/Option");
const pipeable_1 = require("fp-ts/lib/pipeable");
const function_1 = require("fp-ts/lib/function");
const getDateFromTimestamp = (t) => new Date(t);
exports.getDateFromTimestamp = getDateFromTimestamp;
const isNotSilent = (opts) => (_) => !opts.silent;
exports.isNotSilent = isNotSilent;
const filterResponse = (opts) => (req) => (0, pipeable_1.pipe)(O.fromNullable(opts.filter), O.map(filter => filter(req)), O.getOrElse(() => true));
exports.filterResponse = filterResponse;
const formatTime = (timeInMms) => timeInMms > 1000
    ? `+${timeInMms / 1000}s`
    : `+${timeInMms}ms`;
exports.formatTime = formatTime;
const getTimeDifferenceInMs = (startDate) => new Date().getTime() - startDate.getTime();
exports.getTimeDifferenceInMs = getTimeDifferenceInMs;
exports.factorizeTime = (0, function_1.flow)(exports.getDateFromTimestamp, exports.getTimeDifferenceInMs, exports.formatTime);
